Use new Recent Activity, update Offerings API

pjanik opened the pull request:
This PR replaces Recent Activity page with React-based implementation. PortalPages need to be updated first: https://github.com/concord-consortium/portal-pages/pull/45.

Also, Offering API is slightly extended to make that possible. I've added:
 - total progress of student per sequence or activity
 - last run date
 - detailed progress (list of activities and progress per activity)
 
I've removed a bunch of unnecessary code that I could trace. I think that also some parts of Report::Learner and similar classes might not be used anymore, but it might be easier to trace with some tools that Noah might be using during his demolition work.

scytacki commented on app/models/api/v1/offering.rb:
```diff
@@ -41,16 +43,28 @@ def initialize(student, offering)
+                              }
+                            else
+                              nil
+                            end
```
I find this formating difficult to read. I'll change it to my preference. :)

pjanik commented:
@scytacki, I'm assuming that you've already seen optimizations and external report support.

The last commit refactors Offering API and its specs. The old, depreciated routes just call the new one.

The biggest change is in tests. I think that now this API is pretty well tested. Previous tests were based on mocks and pretty much tested access only. I removed them, as new ones cover that too.

I decided that the structure of the response is tested in `#show` specs, as that's the main part of the API. So, most of the test cases that you described in your previous comment are now in `#show` specs (not `#own`). `#own` specs focus on queries, multiple offerings and class filtering, and they simply compare response format to the response of `#show` (as this one is tested already). I think that way it's more manageable long term and easier to update if we change something in the response format. We will just need to update specs of `#show` and other routes will be automatically tested too.
I've also added some tests of depreciated methods.

pjanik commented:
btw, not a big problem, but I've noticed that sometimes we use `own` and sometimes `mine` to return resources that belong to the currently logged in user (in API_V1). I've used `own` here, but well, perhaps it'd be nice to be consistent and keep using one if possible.
`mine` is used by API v1 classes controller
`own` is used by API v1 materials controller
Any preferences? 😉

scytacki commented:
I started this in slack, but then after looking at the PR I came up with a new idea.  However this is extra work, so I'm not sure if it is worth it.

Here are the requirements:
- A report launched from an offering level button should know which offering this button was associated with even if the report type is teacher or class.  This is so a report can present a UI that opens with info about that offering, but it also provides UI elements to navigate to other a class wide summary or to other offerings in the class.  I think currently this was done by parsing the offering id out of the URL.
- Also a report should know how it is configured in the portal: offering, class, or teacher.

Given that we only have one class level report which we could change to match the API, I'm open to changing the URLs passed to a classwide reports and even changing the responses to those classwide report URLs.

However we do have at least one system that is also using the existing routes and their responses directly: workspace.  This is pulling the offering id out of the `show` URL as well as getting the base API path from this URL and then using it to make other requests. I'm also pretty sure sure there is at least one other system using the existing for_class or for_teacher API endpoints directly.  So these need to continue to exist and return the same results (extra fields in the results should be fine though).

So here is an idea: when a report is launched from an offering level button we always pass it the `show` URL for the offering.  We do this even if the report is configured as a class or teacher level report.  We start augmenting the offering result with a link to the class info URL. So a report looking at the offering response can navigate to get the class info.  And we add a parameter to the URL that launches the report that indicates what type of report the portal is configured as. We are already adding a parameters to pass the URL and also the authentication token to this report URL, so adding the report type here seems to make sense.

One addition is to pass the class_info URL as a parameter too. This way the report can make parallel requests for the current offering and the class_info. 

What is missing is non-deprecated way to get all of the offerings for the class. So it seems to make sense to resurrect the for_class route for this. This could be added to the class info result. This is different than `mine?class_id=...` because an admin or external script can request the class_info.

We still keep the `for_teacher` route around for backward compatibility as well as making it easier to make custom external reports.

Writing this makes me think wonder why use `mine?class_id=...` at all. I guess it is because it reduces the number of controller actions. My preference though would be to drop the `mine` from this. Each controller action can point at the same private controller method which can be tested for all of the conditions. And these main controller actions can be just be tested to make sure they call this private method with the right args.

scytacki commented:
@pjanik are you adding any more to this in order to support the reordering of offerings and locking and unlocking offerings, or is that support already here?

pjanik commented:
@scytacki, it's not supported, but it's also not necessary for Recent Activity page.

I'm adding these features to the separate branch (based on this one, tho):
https://github.com/concord-consortium/rigse/tree/assignment-page
and I'll open a new PR when it's ready.

scytacki commented on app/models/external_report.rb:
```diff
@@ -18,42 +17,31 @@ def options_for_report_type
+    class_id = offering.clazz.id
+    class_offerings_url = CGI.escape(routes.api_v1_offerings_url(class_id: class_id, protocol: protocol, host: host))
+    class_url = CGI.escape(routes.api_v1_class_url(class_id, protocol: protocol, host: host))
+    "#{url}?reportType=offering&offering=#{offering_url}&classOfferings=#{class_offerings_url}&class=#{class_url}&token=#{token}&username=#{username}"
```
The URL might already have parameters in it. So it would be better to check for that and not add a second `?` if so.  This code does this: https://github.com/concord-consortium/rigse/blob/6a1f27d36b149612ce90c407f6fd9c0b8bccee08/app/controllers/application_controller.rb#L294-L299 I believe it will remove the need for the `CGI.escape` This isn't a regression, so I'm not going to hold up the PR for this.

scytacki commented on app/models/external_report.rb:
```diff
@@ -18,42 +17,31 @@ def options_for_report_type
-    "#{url}?class=#{class_url}&token=#{token}&username=#{username}"
+    class_url = CGI.escape(routes.api_v1_class_url(class_id, protocol: protocol, host: host))
+    class_offerings_url = CGI.escape(routes.api_v1_offerings_url(class_id: class_id, protocol: protocol, host: host))
+    "#{url}?reportType=class&class=#{class_url}&classOfferings=#{class_offerings_url}&token=#{token}&username=#{username}"
```
Repeat of the above comment about handling external report urls that have parameters in them already.

scytacki commented on app/views/home/_recent.html.haml:
```diff
@@ -1,42 +0,0 @@
-                        .report
-                          %span.linkbutton.pie
-                            = report_link_for(clazz_offering, 'Report', 'Run Report')
-                        - if alternate_report_link_for(clazz_offering)
```
The answer was no, but now it is yes, so this has been fixed.

scytacki commented:
👍 Looks good. I left one comment on handling the external report URLs with parameters already in them.  This wasn't handled before, so it isn't a regression. I'm going to merge this without that fix.
